File: ZezeJava/src/main/java/Zeze/Util/PersistentAtomicLong.java
package Zeze.Util;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

public class PersistentAtomicLong {
    private final AtomicLong currentId = new AtomicLong();
    private volatile long allocated;

    private final String FileName;
    private final int AllocateSize;

    private static ConcurrentHashMap<String, PersistentAtomicLong> pals = new ConcurrentHashMap<>();

    /**
     * 【小优化】请保存返回值，重复使用。
     *
     * @param ProgramInstanceName
     * 程序实例名字。
     * 当进程只有一份实例时，可以直接使用程序名。
     * 有多个实例时，需要另一个Id区分，这里不能使用进程id（pid），需要稳定的。
     * 对于网络程序，可以使用"进程名+Main.Acceptor.Name"
     */
    public static PersistentAtomicLong getOrAdd(final String ProgramInstanceName, final int allocateSize) {
        var name = ProgramInstanceName.replace(':', '.');
        // 这样写，不小心重名也能工作。
        return pals.computeIfAbsent(name, (k) -> new PersistentAtomicLong(k, allocateSize));
    }

    public static PersistentAtomicLong getOrAdd(final String ProgramInstanceName) {
        return getOrAdd(ProgramInstanceName, 5000);
    }

    private PersistentAtomicLong(String ProgramInstanceName, int allocateSize) {
        if (allocateSize <= 0)
            throw new IllegalArgumentException();

        FileName = ProgramInstanceName + ".PersistentAtomicLong.txt";
        AllocateSize = allocateSize;

        if (Files.exists(Path.of(FileName))) {
            try (var r = new BufferedReader(new FileReader(FileName))) {
                var line = r.readLine();
                if (null != line) {
                    allocated = Long.parseLong(line);
                    currentId.set(allocated);
                }
            } catch (IOException ex) {
                throw new RuntimeException(ex);
            }
        }
        // 初始化的时候不allocate，如果程序启动，没有分配就退出，保持原来的值。
    }

    public long next() {
        for (;;) {
            long current = currentId.get();
            if (current >= allocated) {
                allocate();
                continue;
            }
            if (currentId.compareAndSet(current, current + 1))
                return current + 1;
        }
    }

    private void allocate() {
        synchronized (this) {
            if (currentId.get() < allocated)
                return; // has allocated. concurrent.

            allocated += AllocateSize;
            // 应该尽量减少allocate的次数，所以这里文件就不保持打开了。
            try (var pw = new FileWriter(FileName)) {
                pw.write(String.valueOf(allocated));
            }
            catch (IOException ex) {
                throw new RuntimeException(ex);
            }
        }
    }
}

File: Zeze/Util/PersistentAtomicLong.cs

using System.Collections.Concurrent;
using System;
using System.IO;
using System.Text;
using System.Threading;

namespace Zeze.Util
{
    public class PersistentAtomicLong
    {
        private readonly AtomicLong CurrentId = new AtomicLong();
        private AtomicLong allocated = new AtomicLong(); // for volatile long

        public string FileName { get; }
        private readonly int AllocateSize;
        private Mutex Mutex;

        private static ConcurrentDictionary<string, PersistentAtomicLong> pals = new ConcurrentDictionary<string, PersistentAtomicLong>();

        /**
         * 【小优化】请保存返回值，重复使用。
         *
         * @param ProgramInstanceName
         * 程序实例名字。
         * 当进程只有一份实例时，可以直接使用程序名。
         * 有多个实例时，需要另一个Id区分，这里不能使用进程id（pid），需要稳定的。
         * 对于网络程序，可以使用"进程名+Main.Acceptor.Name"
         */
        public static PersistentAtomicLong GetOrAdd(string ProgramInstanceName, int allocateSize)
        {
            var name = ProgramInstanceName.Replace(':', '.');
            // 这样写，不小心重名也能工作。
            return pals.GetOrAdd(name, (k) => new PersistentAtomicLong(k, allocateSize));
        }

        public static PersistentAtomicLong GetOrAdd(string ProgramInstanceName)
        {
            return GetOrAdd(ProgramInstanceName, 5000);
        }

        private PersistentAtomicLong(string ProgramInstanceName, int allocateSize)
        {
            if (allocateSize <= 0)
                throw new ArgumentException();
            FileName = ProgramInstanceName + ".zeze.pal";
            Mutex = new Mutex(false, FileName);
            AllocateSize = allocateSize;

            if (File.Exists(FileName))
            {
                var lines = File.ReadAllLines(FileName);
                if (lines.Length > 0)
                {
                    var alloc = long.Parse(lines[0]);
                    allocated.GetAndSet(alloc);
                    CurrentId.GetAndSet(alloc);
                }
            }
            // 初始化的时候不allocate，如果程序启动，没有分配就退出，保持原来的值。
        }

        public long Next()
        {
            for (; ; )
            {
                long current = CurrentId.Get();
                if (current >= allocated.Get())
                {
                    Allocate();
                    continue;
                }
                if (CurrentId.CompareAndSet(current, current + 1))
                    return current + 1;
            }
        }

        [System.Diagnostics.CodeAnalysis.SuppressMessage("Interoperability", "CA1416:Validate platform compatibility", Justification = "<Pending>")]
        private void Allocate()
        {
            Mutex.WaitOne();
            try
            {
                if (CurrentId.Get() < allocated.Get())
                    return; // has allocated. concurrent.

                // 应该尽量减少allocate的次数，所以这里文件就不保持打开了。
                using var fs = new FileStream(FileName, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.ReadWrite);
                var buffer = new byte[100];
                var rc = fs.Read(buffer);
                var last = (rc <= 0) ? 0 : long.Parse(Encoding.UTF8.GetString(buffer, 0, rc));
                var newlast = last + AllocateSize;
                var reset = newlast < 0;
                if (reset)
                    newlast = AllocateSize;

                fs.SetLength(0);
                fs.Write(Encoding.UTF8.GetBytes(newlast.ToString()));
                allocated.GetAndSet(newlast); // first
                if (reset)
                    CurrentId.GetAndSet(0); // second
            }
            finally
            {
                Mutex.ReleaseMutex();
            }
        }
    }
}
